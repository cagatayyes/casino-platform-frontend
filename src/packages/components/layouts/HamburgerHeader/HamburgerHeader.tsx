import { JSX, useEffect, useState } from 'react';
import { useDatePattern } from 'casino/hooks/useDatePattern';
import { useTranslation } from 'react-i18next';
import { ExitGameModal } from 'casino/components/ExitGameModal/ExitGameModal';
import { useAPI, useScriptAttribute } from 'casino/hooks';
import { SERVER_TIME_ENDPOINT } from 'casino/constants';

interface IHamburgerHeaderProps {
	maxWidth: string;
	isVisible: boolean;
	toggleHamburgerModal: (isVisible: boolean) => void;
	showExitModal:boolean;
	hideExitModal: (isExit: boolean) => void;
}

const HamburgerHeader = ({ maxWidth, isVisible, toggleHamburgerModal, showExitModal, hideExitModal }: IHamburgerHeaderProps): JSX.Element => {
	const [isExitGame, setIsExitGame] = useState<boolean>(showExitModal ?? false);
	const { data: serverTimeData, isLoading, error } = useAPI<number>({ endpoint: SERVER_TIME_ENDPOINT });
	const { datePattern } = useDatePattern(new Date(serverTimeData?.data as number * 1000), 'hh:mm');
	const { t } = useTranslation();
	const { scriptAttributes } = useScriptAttribute();
	const { clientlogo } = scriptAttributes;

	const handleExitGameModal = (): void => {
		setIsExitGame(true);
	};

	const closeExistGameModal = (): void => {
		setIsExitGame(false);
		hideExitModal(false);
	};

	useEffect(() =>{
		if (showExitModal)
			setIsExitGame(showExitModal);

	}, [showExitModal]);

	return (
		<div
			className='absolute top-0 flex flex-row bg-[#0D1335] items-center h-[40px] w-full p-[5px] z-[2001]'
			style={{ maxWidth }}
		>
			<button
				className='flex h-[16px]'
				onClick={() => toggleHamburgerModal(!isVisible)}>
				<span
					className="w-[16px] h-[12px] material-symbols-outlined mt-[2px]"
					style={{ color: isVisible ? '#00D609' : '#fff', fontSize: '16px' }}>
					{isVisible ? 'close' : 'menu'}
				</span>
			</button>
			<div>
				<img
					className='h-33 w-76'
					src={clientlogo}
					width='76'
					height='33'
					alt='Client Logo'
				/>
			</div>
			<div className='flex flow-row items-center w-full justify-end text-right text-[#808594] text-[8px]'>
				{
					!error && !isLoading && datePattern &&
					<span className='font-normal'>{t('Server Time')} {datePattern} | </span>
				}
				<div onClick={handleExitGameModal}
					className='flex flex-row text-[white]'>
					<span className='text-[10px] ml-[3px] mr-[3px] font-bold'>{t('Exit Game')} </span>
					<span
						className='w-[12px] h-[12px] mt-[1px] material-symbols-outlined'
						style={{ fontSize: '12px' }}>
						move_item
					</span>
				</div>
			</div>
			<ExitGameModal
				showExitGameModal={isExitGame}
				closeExitGamemodal={closeExistGameModal}
			/>
		</div>
	);
};

export default HamburgerHeader;
