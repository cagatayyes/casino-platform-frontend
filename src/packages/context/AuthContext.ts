import { createContext } from 'react';

// eslint-disable-next-line @typescript-eslint/naming-convention
interface IAuthContext {
    authToken: string;
    setAuthToken: (authToken: string) => void;
}

export const AuthContext = createContext<IAuthContext>({
	authToken: '',
	setAuthToken: () => { }
});
