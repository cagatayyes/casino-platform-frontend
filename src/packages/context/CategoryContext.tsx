import { CATEGORIES_ENDPOINT } from 'casino/constants';
import { useAPI } from 'casino/hooks';
import { ICategory, ICategoryResponse } from 'casino/types';
import React, { createContext } from 'react';

export const CategoryContext = createContext<ICategoryResponse | undefined>(undefined);

export const CategoryProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
	const { fetchData: fetchCategoryData, data: categoryList, isLoading: isCategoryLoading, error: categoryFetchError } =
		useAPI<ICategory[]>({ endpoint: CATEGORIES_ENDPOINT });

	return (
		<CategoryContext.Provider value={{ categoryList, isCategoryLoading, categoryFetchError, fetchCategoryData }}>
			{children}
		</CategoryContext.Provider>
	);
};

