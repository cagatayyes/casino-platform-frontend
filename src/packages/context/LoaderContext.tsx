import { Loader } from 'casino/components';
import { createContext, useState } from 'react';

export interface ILoaderContext {
	isLoading: boolean;
	showLoader: (loaderMessage: string) => void;
	hideLoader: () => void;
}

export const LoaderContext = createContext<ILoaderContext | undefined>(undefined);

export const LoaderProvider = ({ children }: {children: React.ReactNode }): JSX.Element => {
	const [isLoading, setIsLoading] = useState(false);
	const [loadingText, setLoadingText] = useState('');

	const showLoader = (loaderMessage: string): void => {
		setIsLoading(true);
		setLoadingText(loaderMessage);
	};
	const hideLoader = (): void => setIsLoading(false);

	return (
		<LoaderContext.Provider value={{ isLoading, showLoader, hideLoader }}>
			<>
				{isLoading && <Loader loaderText={loadingText} />}
				{children}
			</>
		</LoaderContext.Provider>
	);
};
