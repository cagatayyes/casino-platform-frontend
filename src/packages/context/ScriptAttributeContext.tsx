import { CASINO_CLIENT_LOGO, CASINO_NOT_FOUND, CASINO_WIDGET_SCRIPT_ID } from 'casino/constants';
import { ICasinoScriptAttribute, IScriptAttribute, SCRIPT_ATTRIBUTE_KEYS } from 'casino/types';
import React, { Dispatch, createContext, useEffect, useReducer } from 'react';

export const scriptAttributeInitialState: IScriptAttribute = {
	token: '',
	currency: 'NGN',
	maxwidth: '640px',
	timezone: 'Africa/Lagos',
	clientlogo: CASINO_CLIENT_LOGO,
	pagenotfoundlogo: CASINO_NOT_FOUND,
	isTokenChanged: false,
	baseURL: '/casino'
};

type Action = | { type: 'SET_SCRIPT_ATTRIBUTE', payload: IScriptAttribute }
	| { type: 'SET_TOKEN_CHANGED', payload: boolean };

const scriptAttributeReducer = (state: IScriptAttribute, action: Action): IScriptAttribute => {
	switch (action.type) {
	case 'SET_SCRIPT_ATTRIBUTE': {
		const isTokenChanged = action.payload.token !== state.token;

		return { ...state, ...action.payload, isTokenChanged };
	}
	case 'SET_TOKEN_CHANGED': {
		return { ...state, isTokenChanged: action.payload };
	}
	default:
		return state;
	}
};

export interface IScriptAttributeContext {
    scriptAttributes: IScriptAttribute,
	scriptAttributesDispatch: Dispatch<Action>;
}

const getScriptAttributes = (): IScriptAttribute => {
	const casinoScriptAttributes: NamedNodeMap | undefined = document.getElementById(CASINO_WIDGET_SCRIPT_ID)?.attributes;

	return Array.from(casinoScriptAttributes || []).reduce((attribute: ICasinoScriptAttribute, { name, value }) => {
		if (SCRIPT_ATTRIBUTE_KEYS.includes(name)) {
			attribute[name] = value;
		}

		return attribute;
	}, { ...scriptAttributeInitialState });
};

const observeScriptAttributes = (scriptAttributesDispatch: Dispatch<Action>): (() => void) => {
	const casinoWidgetScript = document.getElementById(CASINO_WIDGET_SCRIPT_ID);
	if (!casinoWidgetScript) return () => {};

	const config = { attributes: true };
	const mutationObserverCallback = (mutationsList: MutationRecord[]): void => {
		for (const mutation of mutationsList) {
			if (mutation.type === 'attributes' && mutation.attributeName) {
				const name: string = mutation.attributeName;
				const newValue: string | null = casinoWidgetScript.getAttribute(name);

				if (SCRIPT_ATTRIBUTE_KEYS.includes(name)) {
					scriptAttributesDispatch({
						type: 'SET_SCRIPT_ATTRIBUTE',
						payload: { [name]: newValue } as ICasinoScriptAttribute,
					});
				}
			}
		}
	};

	const observer = new MutationObserver(mutationObserverCallback);
	observer.observe(casinoWidgetScript, config);

	return (): void => observer.disconnect();
};

export const ScriptAttributeContext = createContext<IScriptAttributeContext | undefined>(undefined);

export const ScriptAttributeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
	const initialState = getScriptAttributes();
	const [scriptAttributes, scriptAttributesDispatch] = useReducer(scriptAttributeReducer, initialState);

	useEffect(() => {
		const casinoWidgetObserver =  observeScriptAttributes(scriptAttributesDispatch);

		return () => {
			casinoWidgetObserver();
		};
	}, []);

	return (
		<ScriptAttributeContext.Provider value={{ scriptAttributes, scriptAttributesDispatch }}>
			{children}
		</ScriptAttributeContext.Provider>
	);
};
