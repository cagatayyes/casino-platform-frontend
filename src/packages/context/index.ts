export * from './LoaderContext';
export * from './AuthContext';
export * from './ScriptAttributeContext';
export * from './CategoryContext';
