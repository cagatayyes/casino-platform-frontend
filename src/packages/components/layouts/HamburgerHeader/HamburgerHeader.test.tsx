import { render } from '@testing-library/react';
import HamburgerHeader from './HamburgerHeader';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { ScriptAttributeProvider } from '../../../context/ScriptAttributeContext';

jest.mock('casino/hooks/useDatePattern', () => ({
	// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
	useDatePattern: () => ({
		datePattern: '1708591412' // Mocked date pattern
	})
}));

/* eslint-disable react/prop-types */
jest.mock('swiper/react', () => ({
	Swiper: ({ children }): JSX.Element => <div>{children}</div>,
	SwiperSlide: ({ children }): JSX.Element => <div>{children}</div>
}));

jest.mock('swiper/modules', () => ({
	Swiper: ({ children }): JSX.Element => <div>{children}</div>,
	SwiperSlide: ({ children }): JSX.Element => <div>{children}</div>
}));

describe('Hamburger Header', () => {
	it('should render Hamburger Header navigation', () => {
		const component = render(
			<BrowserRouter>
				<ScriptAttributeProvider>
					<HamburgerHeader
						maxWidth='640px'
						isVisible={false}
						toggleHamburgerModal={jest.fn()}
						showExitModal={false}
						hideExitModal={jest.fn()}
					/>
				</ScriptAttributeProvider>
			</BrowserRouter>
		);
		expect(component).toMatchSnapshot();
	});
});
