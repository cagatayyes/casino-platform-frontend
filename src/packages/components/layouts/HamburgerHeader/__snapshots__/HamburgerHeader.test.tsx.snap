// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hamburger Header should render Hamburger Header navigation 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="absolute top-0 flex flex-row bg-[#0D1335] items-center h-[40px] w-full p-[5px] z-[2001]"
        style="max-width: 640px;"
      >
        <button
          class="flex h-[16px]"
        >
          <span
            class="w-[16px] h-[12px] material-symbols-outlined mt-[2px]"
            style="color: rgb(255, 255, 255); font-size: 16px;"
          >
            menu
          </span>
        </button>
        <div>
          <img
            alt="Client Logo"
            class="h-33 w-76"
            height="33"
            src="https://b2b-demo-development.1337pro.com/static/media/logo.81d2c3cf27aa6d2aaffe8ab0acf4bbab.svg"
            width="76"
          />
        </div>
        <div
          class="flex flow-row items-center w-full justify-end text-right text-[#808594] text-[8px]"
        >
          <div
            class="flex flex-row text-[white]"
          >
            <span
              class="text-[10px] ml-[3px] mr-[3px] font-bold"
            >
              Exit Game
               
            </span>
            <span
              class="w-[12px] h-[12px] mt-[1px] material-symbols-outlined"
              style="font-size: 12px;"
            >
              move_item
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class=""
      id="casino-modal-container"
    />
  </body>,
  "container": <div>
    <div
      class="absolute top-0 flex flex-row bg-[#0D1335] items-center h-[40px] w-full p-[5px] z-[2001]"
      style="max-width: 640px;"
    >
      <button
        class="flex h-[16px]"
      >
        <span
          class="w-[16px] h-[12px] material-symbols-outlined mt-[2px]"
          style="color: rgb(255, 255, 255); font-size: 16px;"
        >
          menu
        </span>
      </button>
      <div>
        <img
          alt="Client Logo"
          class="h-33 w-76"
          height="33"
          src="https://b2b-demo-development.1337pro.com/static/media/logo.81d2c3cf27aa6d2aaffe8ab0acf4bbab.svg"
          width="76"
        />
      </div>
      <div
        class="flex flow-row items-center w-full justify-end text-right text-[#808594] text-[8px]"
      >
        <div
          class="flex flex-row text-[white]"
        >
          <span
            class="text-[10px] ml-[3px] mr-[3px] font-bold"
          >
            Exit Game
             
          </span>
          <span
            class="w-[12px] h-[12px] mt-[1px] material-symbols-outlined"
            style="font-size: 12px;"
          >
            move_item
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
